### PR TITLE
[pipeline] CogVideoX-Fun Control

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Adds support for CogVideoX-Fun Pose Control models

- [`alibaba-pai/CogVideoX-Fun-V1.1-2b-Pose`](https://huggingface.co/alibaba-pai/CogVideoX-Fun-V1.1-2b-Pose)
- [`alibaba-pai/CogVideoX-Fun-V1.1-5b-Pose`](https://huggingface.co/alibaba-pai/CogVideoX-Fun-V1.1-5b-Pose)

### Usage

```python
import torch
from diffusers import CogVideoXFunControlPipeline, DDIMScheduler
from diffusers.utils import export_to_video, load_video

pipe = CogVideoXFunControlPipeline.from_pretrained(
    "alibaba-pai/CogVideoX-Fun-V1.1-5b-Pose", torch_dtype=torch.bfloat16
)
pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
pipe.to("cuda")

control_video = load_video(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/hiker.mp4"
)
prompt = (
    "An astronaut stands triumphantly at the peak of a towering mountain. Panorama of rugged peaks and "
    "valleys. Very futuristic vibe and animated aesthetic. Highlights of purple and golden colors in "
    "the scene. The sky is looks like an animated/cartoonish dream of galaxies, nebulae, stars, planets, "
    "moons, but the remainder of the scene is mostly realistic."
)

video = pipe(prompt=prompt, control_video=control_video).frames[0]
export_to_video(video, "output.mp4", fps=8)
```

<table align=center>
<tr>
  <td align=center><video src="https://github.com/user-attachments/assets/311a04fa-c9ec-4838-848a-c00ed59f7dbf"> Your browser does not support the video tag. </video></td>
  <td align=center><video src="https://github.com/user-attachments/assets/4f5251e8-cdc0-4d90-9d0a-d35f51810268"> Your browser does not support the video tag. </video></td>
</tr>
</table>

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@bubbliiiing @yunkchen

### CODE REVIEW

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+                if use_dynamic_cfg:
+                    self._guidance_scale = 1 + guidance_scale * (
+                        (1 - math.cos(math.pi * ((num_inference_steps - t.item()) / num_inference_steps) ** 5.0)) / 2
+                    )
+                if do_classifier_free_guidance:
+                    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
+                    noise_pred = noise_pred_uncond + self.guidance_scale * (noise_pred_text - noise_pred_uncond)
+
+                # compute the previous noisy sample x_t -> x_t-1
+                if not isinstance(self.scheduler, CogVideoXDPMScheduler):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py:764]: if it works with ddim, I think it works with all other regular schedulers, let's try out with a couple non-cogvideox schedulers, and maybe just remove these lines

**Review by yiyixuxu — COMMENTED** [review]
thanks! awesome! so we did not need to update the model at all?
I left comments about scheduler

**Comment by a-r-r-o-w** [comment]
> so we did not need to update the model at all?

nope, no changes at all :)

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py (hunk 1 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py:764]: Yep, already did and it does work with the common ones. Will remove

Comment (a-r-r-o-w) [reply src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py]: So, CogVideoX-Fun does not work with the CogVideoX-specific schedulers we've added but works well with most of the common ones I tried. From their training scripts, they are indeed training on a different objective, so it makes sense. Updated to remove this

**Comment by bubbliiiing** [comment]
Thank you for the cc. How can I help you?

**Comment by a-r-r-o-w** [comment]
Hi @bubbliiiing. Thanks for your awesome work! I would be grateful if you could give this PR a review and verify its correctness and if it's compatible with the original pipeline/scripts. Happy to merge this if you think it's good (and subsequently the inpainting models in a different PR) :)

Also would love to collaborate on any future releases you are planning on and if we can help with training/finetuning in any way as @sayakpaul and I are looking into it too alongside the CogVideoX team

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+            width (`int`, *optional*, defaults to self.transformer.config.sample_height * self.vae_scale_factor_spatial):
+                The width in pixels of the generated image. This is set to 720 by default for the best results.
+            num_inference_steps (`int`, *optional*, defaults to 50):
+                The number of denoising steps. More denoising steps usually lead to a higher quality image at the
+                expense of slower inference.
+            timesteps (`List[int]`, *optional*):
+                Custom timesteps to use for the denoising process with schedulers which support a `timesteps` argument
+                in their `set_timesteps` method. If not defined, the default behavior when `num_inference_steps` is
+                passed will be used. Must be in descending order.
+            guidance_scale (`float`, *optional*, defaults to 7.0):
```
Comment (stevhliu) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py:577]:
```suggestion
            guidance_scale (`float`, *optional*, defaults to 6.0):
```

File: docs/source/en/api/pipelines/cogvideox.md [hunk 3]
```diff
@@ -36,6 +36,10 @@ There are two models available that can be used with the text-to-video and video
 There is one model available that can be used with the image-to-video CogVideoX pipeline:
 - [`THUDM/CogVideoX-5b-I2V`](https://huggingface.co/THUDM/CogVideoX-5b-I2V): The recommended dtype for running this model is `bf16`.
 
+There are two models that allow controllable generation (by the [Alibaba-PAI](https://huggingface.co/alibaba-pai) team):
```
Comment (stevhliu) [inline docs/source/en/api/pipelines/cogvideox.md:39]:
```suggestion
There are two models that support pose controllable generation (by the [Alibaba-PAI](https://huggingface.co/alibaba-pai) team):
```

**Review by stevhliu — APPROVED** [review]
LGTM, thanks!

**Review by yiyixuxu — APPROVED** [review]
awesome!

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+        if latents is None:
+            latents = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            latents = latents.to(device)
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    def prepare_control_latents(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py:356]: We could maybe provide a reference we used to implement the `prepare_control_latents()` function if it's not fairly generic.

**Review by sayakpaul — APPROVED** [review]
Awesome. Just a single a comment!

**Review by bubbliiiing — APPROVED** [review]
LGTM.

**Comment by bubbliiiing** [comment]
This is the test result after I simply modified the prompt words, it is very good. 
https://github.com/user-attachments/assets/daddc0eb-e2c4-44ac-b86e-24a668308de9

**Comment by sayakpaul** [comment]
@bubbliiiing thanks for reviewing and for testing this PR. 

Would it be possible to also add a note about `diffusers` support in your repo?

**Comment by bubbliiiing** [comment]
> @bubbliiiing thanks for reviewing and for testing this PR.
> 
> Would it be possible to also add a note about `diffusers` support in your repo?

Of course! Could you provide me with an example of adding "diffusers support"? I would like to add it in the same way.

**Comment by sayakpaul** [comment]
> Of course! Could you provide me with an example of adding "diffusers support"? I would like to add it in the same way.

Maybe something like:

> CongVideoX-Fun Control is now supported in `diffusers`. Thanks to [a-r-r-o-w](https://github.com/a-r-r-o-w) who contributed the support in [this PR](https://github.com/huggingface/diffusers/pull/9671). Check out the [docs](https://huggingface.co/docs/diffusers/main/en/api/pipelines/cogvideox) to know more.

**Comment by bubbliiiing** [comment]
> > Of course! Could you provide me with an example of adding "diffusers support"? I would like to add it in the same way.
> 
> Maybe something like:
> 
> > CongVideoX-Fun Control is now supported in `diffusers`. Thanks to [a-r-r-o-w](https://github.com/a-r-r-o-w) who contributed the support in [this PR](https://github.com/huggingface/diffusers/pull/9671). Check out the [docs](https://huggingface.co/docs/diffusers/main/en/api/pipelines/cogvideox) to know more.

I understand. I'll submit a PR to the CogVideoX-Fun repository.

**Comment by bubbliiiing** [comment]
Mission completed. image

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py (hunk 4 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_control.py:357]: added
